### ISSUE TITLE
DbException not properly translated

### ISSUE THREAD

**Issue by mp911de** [r2dbc/r2dbc-h2#83]
Exceptions raised by H2 are `DbException` (runtime exceptions). These are not properly translated into R2DBC exceptions.

Reproducer:

```java
H2ConnectionFactory connectionFactory = new H2ConnectionFactory(H2ConnectionConfiguration
		.builder() //
		.inMemory("r2dbc") //
		.username("sa") //
		.password("") //
		.option("DB_CLOSE_DELAY=-1").build());

Flux.from(connectionFactory.create())
		.flatMap(h2Connection -> h2Connection
				.createStatement("SELECT foobar FROM unknown").execute())
		.blockLast();
```

Stack trace:

```
Exception in thread "main" org.h2.message.DbException: Tabelle "UNKNOWN" nicht gefunden
Table "UNKNOWN" not found [42102-199]
	at org.h2.message.DbException.get(DbException.java:205)
	at org.h2.message.DbException.get(DbException.java:181)
	at org.h2.command.Parser.readTableOrView(Parser.java:7146)
	at org.h2.command.Parser.readTableFilter(Parser.java:1895)
	at org.h2.command.Parser.parseSelectSimpleFromPart(Parser.java:2641)
	at org.h2.command.Parser.parseSelectSimple(Parser.java:2788)
	at org.h2.command.Parser.parseSelectSub(Parser.java:2636)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:2469)
	at org.h2.command.Parser.parseSelect(Parser.java:2440)
	at org.h2.command.Parser.parsePrepared(Parser.java:814)
	at org.h2.command.Parser.parse(Parser.java:788)
	at org.h2.command.Parser.parse(Parser.java:760)
	at org.h2.command.Parser.prepareCommand(Parser.java:683)
	at org.h2.engine.Session.prepareLocal(Session.java:627)
	at org.h2.engine.Session.prepareCommand(Session.java:565)
	at io.r2dbc.h2.client.SessionClient.createCommand(SessionClient.java:117)
	at io.r2dbc.h2.client.SessionClient.lambda$prepareCommand$3(SessionClient.java:93)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:100)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1575)
	at reactor.core.publisher.FluxDefaultIfEmpty$DefaultIfEmptySubscriber.onComplete(FluxDefaultIfEmpty.java:100)
	at reactor.core.publisher.Operators.complete(Operators.java:132)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:122)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:63)
	at reactor.core.publisher.FluxDefaultIfEmpty.subscribe(FluxDefaultIfEmpty.java:42)
	at reactor.core.publisher.FluxMap.subscribe(FluxMap.java:62)
	at reactor.core.publisher.FluxPeek.subscribe(FluxPeek.java:83)
	at reactor.core.publisher.FluxFlatMap.subscribe(FluxFlatMap.java:97)
	at reactor.core.publisher.Flux.subscribe(Flux.java:7922)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:389)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
	at reactor.core.publisher.FluxJust$WeakScalarSubscription.request(FluxJust.java:99)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:335)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90)
	at reactor.core.publisher.FluxJust.subscribe(FluxJust.java:70)
	at reactor.core.publisher.FluxMapFuseable.subscribe(FluxMapFuseable.java:63)
	at reactor.core.publisher.FluxFlatMap.subscribe(FluxFlatMap.java:97)
	at reactor.core.publisher.Flux.subscribe(Flux.java:7922)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:389)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2131)
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:155)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:335)
	at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:86)
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:54)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
	at reactor.core.publisher.MonoMap.subscribe(MonoMap.java:55)
	at reactor.core.publisher.FluxSourceMono.subscribe(FluxSourceMono.java:46)
	at reactor.core.publisher.FluxFlatMap.subscribe(FluxFlatMap.java:97)
	at reactor.core.publisher.Flux.blockLast(Flux.java:2389)
	at org.springframework.data.r2dbc.testing.H2TestSupport.main(H2TestSupport.java:84)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:93)
		at reactor.core.publisher.Flux.blockLast(Flux.java:2390)
		... 1 more
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Tabelle "UNKNOWN" nicht gefunden
Table "UNKNOWN" not found; SQL statement:
SELECT foobar FROM unknown [42102-199]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:451)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:427)
	... 51 more
```

Should be:

`R2dbcBadGrammarException`

Looking at the code, the mock setup does not match how `SessionClient` behaves.

**Comment by gregturn** [comment]
Hmm. Based on #69, we moved away from translating internal H2 exceptions to instead base things on JDBC exceptions.

Appears that thinking was flawed. Perhaps our usage of H2 circumvents the exception translation layer H2 itself does to convert to JDBC exceptions.

Thoughts @katzyn?

**Comment by katzyn** [comment]
When you use internals of H2 you can get a `DbException` that needs to be converted to a normal `SQLException` with `DbException.toSQLException(...)`.

**Comment by katzyn** [comment]
It's not related, but `UNKNOWN` will be a keyword in the next version of H2. If you use it in your unit tests it's better to replace it with something else, otherwise error codes and messages may be different after upgrade.

**Comment by mp911de** [comment]
Exception translation to `SQLException` happens in H2 in the JDBC facade itself. We should follow the same design pattern to introduce an exception mapper and translate H2 exception to R2DBC exceptions in our R2DBC facade.

**Comment by gregturn** [comment]
Agreed.

We have https://github.com/r2dbc/r2dbc-h2/blob/master/src/main/java/io/r2dbc/h2/H2DatabaseExceptionFactory.java. I just need to rewire it into the now imperative code performance enhancement, using @katzyn 's suggested converter utlity.

**Comment by gregturn** [comment]
@mp911de If you want to check it out, https://github.com/r2dbc/r2dbc-h2/tree/exceptions supposedly has proper support now for `DbException` translation in place.

I need to draft more test cases to really verify this is covered top to bottom, but you're free to start checking against Spring Data R2dbc at the same time.

**Comment by mp911de** [comment]
That looks quite good. In Spring Data R2DBC we have one or two tests for exceptions, though. We likely just need to catch `DbException`, `SQLException` isn't thrown from anywhere.

**Comment by gregturn** [comment]
Resolved via 2dc8cac.
